tests/KubernetesClient.Tests/PipeStream.cs

```diff
@@ -0,0 +1,413 @@
+            }
+            DataAvailable -= read;
+            return read;
+        }
+
+        /// <summary>Writes data into the buffer. This method will not block.</summary>
+        public void Write(byte[] data, int offset, int count) => Write(new ReadOnlySpan<byte>(data, offset, count));
+
+        /// <summary>Writes data into the buffer. This method will not block.</summary>
+        public void Write(ReadOnlySpan<byte> data)
+        {
+            int length = data.Length;
+            if(length != 0)
+            {
+                if(dataList.Count == 0) dataList.AddLast(new Chunk(length)); // if the chunk list is empty, add a chunk
+                do // while there's still data left to write...
+                {
+                    Chunk chunk = dataList.Last.Value; // get the last chunk
+                    if(chunk.AvailableSpace == 0) // if it's full...
+                    {
+                        chunk = new Chunk(data.Length); // add a new chunk
+                        dataList.AddLast(chunk);
+                    }
+                    int toCopy = Math.Min(chunk.AvailableSpace, data.Length); // now copy as much as we can into the chunk
+                    chunk.Write(data.Slice(0, toCopy));
+                    data = data.Slice(toCopy);
+                } while(data.Length != 0);
+                DataAvailable += length;
+            }
+        }
+
+        const int MinChunkSize = 32*1024, DropThreshold = 64*1024;
+
+        #region Chunk
+        /// <summary>Represents a chunk of data. We maintain multiple chunks of data rather than a single large buffer to avoid having
+        /// to reallocate arrays or copy data.
+        /// </summary>
+        sealed class Chunk
+        {
+            public Chunk(int minSize) => Buffer = ArrayPool<byte>.Shared.Rent(Math.Max(MinChunkSize, minSize));
+            public int AvailableData => WriteIndex - ReadIndex;
+            public int AvailableSpace => Buffer.Length - WriteIndex;
+            public int ReadIndex { get; private set; }
+            public int WriteIndex { get; private set; }
+
+            public void Dispose() => ArrayPool<byte>.Shared.Return(Buffer);
+
+            public void Read(int byteCount, Span<byte> dest)
+            {
+                new Span<byte>(Buffer, ReadIndex, byteCount).CopyTo(dest);
+                ReadIndex += byteCount;
+            }
+
+            public void Reset() => ReadIndex = WriteIndex = 0;
+
+            public void Write(ReadOnlySpan<byte> span)
+            {
+                span.CopyTo(new Span<byte>(Buffer, WriteIndex, AvailableSpace));
+                WriteIndex += span.Length;
+            }
+
+            public readonly byte[] Buffer;
+        }
+        #endregion
+
+        readonly LinkedList<Chunk> dataList = new LinkedList<Chunk>();
+    }
+    #endregion
+
+    #region PipeStream
+    /// <summary>Represents a read-only stream backed by a <see cref="Pipe"/>.</summary>
+    sealed class PipeStream : Stream
```

Comment:
A memory stream wouldn't work the same way. Instead of blocking at the end of it, it will return EOF. And it's not really suitable for simultaneous reading and writing.